src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -111,231 +243,217 @@ def compute_tof_lookup_table(
+    )
 
-def unwrapped_time_of_arrival(
-    da: RawData, offset: PulseStrideOffset, pulse_period: PulsePeriod
-) -> UnwrappedTimeOfArrival:
-    """
-    Compute the unwrapped time of arrival of the neutron at the detector.
-    For event data, this is essentially ``event_time_offset + event_time_zero``.
+    # Compute time-of-flight of the bin edges using the interpolator
+    tofs = sc.array(
+        dims=etos.dims,
+        values=interp((pulse_index.values, ltotal.values, etos.values)),
+        unit=eto_unit,
+    )
 
-    Parameters
-    ----------
-    da:
-        Raw detector data loaded from a NeXus file, e.g., NXdetector containing
-        NXevent_data.
-    offset:
-        Integer offset of the first pulse in the stride (typically zero unless we are
-        using pulse-skipping and the events do not begin with the first pulse in the
-        stride).
-    pulse_period:
-        Period of the source pulses, i.e., time between consecutive pulse starts.
-    """
-    if da.bins is None:
-        # 'time_of_flight' is the canonical name in NXmonitor, but in some files, it
-        # may be called 'tof'.
-        key = next(iter(set(da.coords.keys()) & {"time_of_flight", "tof"}))
-        toa = da.coords[key]
-    else:
-        # To unwrap the time of arrival, we want to add the event_time_zero to the
-        # event_time_offset. However, we do not really care about the exact datetimes,
-        # we just want to know the offsets with respect to the start of the run.
-        # Hence we use the smallest event_time_zero as the time origin.
-        time_zero = da.coords["event_time_zero"] - da.coords["event_time_zero"].min()
-        coord = da.bins.coords["event_time_offset"]
-        unit = elem_unit(coord)
-        toa = (
-            coord
-            + time_zero.to(dtype=float, unit=unit, copy=False)
-            - (offset * pulse_period).to(unit=unit, copy=False)
+    return rebinned.assign_coords(tof=tofs)
+
+
+def _time_of_flight_data_events(
+    da: sc.DataArray,
+    lookup: sc.DataArray,
+    ltotal: sc.Variable,
+    pulse_period: sc.Variable,
+    pulse_stride: int,
+    pulse_stride_offset: int,
+) -> sc.DataArray:
+    etos = da.bins.coords["event_time_offset"]
+    eto_unit = elem_unit(etos)
+    pulse_period = pulse_period.to(unit=eto_unit)
+    frame_period = pulse_period * pulse_stride
+
+    # TODO: Finding the `tmin` below will not work in the case were data is processed
+    # in chunks, as taking the minimum time in each chunk will lead to inconsistent
+    # pulse indices (this will be the case in live data, or when using the
+    # StreamProcessor). We could instead read it from the first chunk and store it?
+
+    # Compute a pulse index for every event: it is the index of the pulse within a
+    # frame period. When there is no pulse skipping, those are all zero. When there is
+    # pulse skipping, the index ranges from zero to pulse_stride - 1.
+    tmin = da.bins.coords['event_time_zero'].min()
```

Comment:
Maybe one could simply use epoch? Or some other fixed datetime?

Reply:
The more I think about this, the more it seems there's something wrong with the approach.


I tried with just using epoch, and for the examples we have, it also works.
But if I offset my event time zeros by a pulse period, the results are all wrong until I correct it using `PulseStrideOffset=1`.

This is reproducing the situation where one would start recording the file one pulse period later.
In practice, we can never really know when this was.
I'm thinking we can only know by looking at the open and close times of the pulse skipping chopper?

Reply:
I have an ugly hack in https://github.com/scipp/essreduce/tree/pulse-index which basically tries all possible `PulseStrideOffset`s and then picks the result with the least number of NaNs in it.

It works for the examples I've tried, but it doesn't feel safe and it also doubles or triples the cost of computing tof.
We could optimize by using just a small number of events at the start of the event buffer to decide which offset to go with, but it doesn't fix the fact that it doesn't feel very robust (might actually make robustness worse; how do you decide when you have enough neutrons to perform the test?)